chainer/initializers/constant.py

```diff
@@ -51,13 +51,16 @@ def __call__(self, array):
         if self.dtype is not None:
             assert array.dtype == self.dtype
 
+        fill_value = self.fill_value
         # Calling copy to ensures that the fill_value array
         # is moved to the device where array resides
-        if isinstance(self.fill_value, chainer.get_array_types()):
-            backend.copyto(array, self.fill_value)
+        if isinstance(fill_value, chainer.get_array_types()):
+            if self.dtype is not None:
+                fill_value = fill_value.astype(self.dtype, copy=False)
+            backend.copyto(array, fill_value)
```

Comment:
I think these 3 lines can be written simply as:
```py
backend.copyto(array, self.fill_value.astype(array.dtype, copy=False))
```

Reply:
It fails for ChainerX. Should I add a TODO comment?

Reply:
How does it fail?

Reply:
```python
chainerx.arange(1).astype(dtype=None) # TypeError: Dtype not understood: None
numpy.arange(1).astype(dtype=None) # array([0.]) with float64 dtype
```

Reply:
Sorry now I noticed you wrote `array.dtype`.

Reply:
Amended changes, PTAL.